File: clj/2024/src/advent/day_01.clj
(ns advent.day-01
  (:require [advent.util :as util]))

(def input-path "resources/day01/")

(defn star-1 [input-file-name]
  (let [lines (util/read-file-into-list-of-lists input-file-name)
        left-list (->> lines (map first) sort (map #(Integer/parseInt %)))
        right-list (->> lines (map second) sort (map #(Integer/parseInt %)))
        paired (map (fn [a b] [a b]) left-list right-list)
        distances (map (fn [[a b]] (Math/abs (- a b))) paired)
        sum (reduce + distances)]
    sum))

(defn star-2 [input-file-name]
  (let [lines (util/read-file-into-list-of-lists input-file-name)
        left-list (->> lines (map first) sort (map #(Integer/parseInt %)))
        right-list (->> lines (map second) sort (map #(Integer/parseInt %)))
        times-in-list-fn (fn [item list-of-items]
                           (reduce (fn [acc curr]
                                     (if (= item curr)
                                       (inc acc)
                                       acc)
                                     ) 0 list-of-items))

        times-in-list-mult (->> left-list (map (fn [item] (* (times-in-list-fn item right-list)
                                                            item))))]
    (reduce + times-in-list-mult)))

(comment
  (star-1 (str input-path "input.txt"))
  (star-2 (str input-path "input.txt")))

File: kotlin/src/main/kotlin/aoc2024/Day01.kt
package aoc2024

import resourceAsListOfString

fun solveDay01Part1(inputPath: String) {
  val lines = resourceAsListOfString(inputPath)

  val (left, right) = lines.map {
    val (a, b) = it.split("\\s+".toRegex()).map { it.toInt() }
    Pair(a, b)
  }.unzip()

  left.sorted()
    .zip(right.sorted())
    .map { Math.abs(it.first - it.second) }
    .sum()
    .also { println(it) }
}

fun solveDay01Part2(inputPath: String) {
  val lines = resourceAsListOfString(inputPath)

  val (left, right) = lines.map {
    val (a, b) = it.split("\\s+".toRegex()).map { it.toInt() }
    Pair(a, b)
  }.unzip()

  left.map { curr -> right.count { curr == it } * curr }
    .sum()
    .also { println(it) }
}


fun main() {

  solveDay01Part1("2024/day01_sample.txt")
  solveDay01Part1("2024/day01.txt")

  solveDay01Part2("2024/day01_sample.txt")
  solveDay01Part2("2024/day01.txt")

}